File: adapter/src/main/java/org/jvalue/ods/adapterservice/interpreter/Interpreter.java
package org.jvalue.ods.adapterservice.interpreter;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.databind.JsonNode;

import java.io.IOException;
import java.util.Map;
import java.util.Map.Entry;

public abstract class Interpreter {

    public abstract String getType();
    public abstract String getDescription();

    @JsonProperty("parameters")
    public abstract Map<String, String> getAvailableParameters();

    public final JsonNode interpret(String data, Map<String, Object> parameters) throws IOException {
      validateParameters(parameters);
      return doInterpret(data, parameters);
    }
    protected abstract JsonNode doInterpret(String data, Map<String, Object> parameters) throws IOException;

    protected void validateParameters(Map<String, Object> inputParameters) {
      boolean illegalArguments = false;
      String illegalArgumentsMessage = "";
      for(Entry<String, String> requiredParameter: getAvailableParameters().entrySet()) {
        if(inputParameters.get(requiredParameter.getKey()) == null) {
          illegalArguments = true;
          illegalArgumentsMessage = illegalArgumentsMessage + getType() + " importer requires parameter " + requiredParameter.getKey() + "/n";
        }
      }
      if(illegalArguments) {
        throw new IllegalArgumentException(illegalArgumentsMessage);
      }
    }
}

File: adapter/src/interpreter/Interpreter.ts
export abstract class Interpreter {
  type: string | undefined;
  description: string | undefined;
  parameters: Record<string, unknown> | undefined;


  interpret(data: string, parameters: Record<string, unknown>): string { //throws IOException
    this.validateParameters(parameters);
    return this.doInterpret(data, parameters);
  }

  abstract doInterpret(data: string, parameters: Record<string, unknown>): string //throws IOException;

  validateParameters(inputParameters: Record<string, unknown>) { //throws InterpreterParameterException
    /*boolean illegalArguments = false;
      String illegalArgumentsMessage = "";

      for (InterpreterParameterDescription requiredParameter : getAvailableParameters()) {
        if (inputParameters.get(requiredParameter.getName()) == null) {
          illegalArguments = true;
          illegalArgumentsMessage = illegalArgumentsMessage + getType() + " interpreter requires parameter "
            + requiredParameter.getName() + "/n";

        } else if (inputParameters.get(requiredParameter.getName()).getClass() != requiredParameter.getType()) {
          illegalArguments = true;
          illegalArgumentsMessage = illegalArgumentsMessage + getType() + " interpreter requires parameter "
            + requiredParameter.getName() + " to be type " + requiredParameter.getType().toString() + "/n";
        }
      }
      if (illegalArguments) {
        throw new InterpreterParameterException(illegalArgumentsMessage);
      }*/
    }
  }
